python/tests/test_tree_stats.py

```diff
@@ -2515,183 +2548,372 @@ class TestSitef4(Testf4, MutatedTopologyExamplesMixin):
+                else:
+                    allele_count[ancestral_state] -= count[mutation.node]
+                mutation_index += 1
+            if polarised:
+                del allele_count[ancestral_state]
+
+            pos = sites.position[site_index]
+            while windows[window_index + 1] <= pos:
+                window_index += 1
+            assert windows[window_index] <= pos < windows[window_index + 1]
+            site_result = result[window_index]
+
+            for allele, c in allele_count.items():
+                # If the allele is not present in any of the sample sets we
+                # do not consider it for the AFS
+                if np.sum(c) > 0:
+                    if polarised:
+                        site_result[tuple(c)] += 1
+                    else:
+                        # Fold the counts
+                        c = [
+                            min(c[k], len(sample_set) - c[k])
+                            for k, sample_set in enumerate(sample_sets)]
+                        site_result[tuple(c)] += 0.5
+            site_index += 1
+
+    if span_normalise:
+        for j in range(num_windows):
+            span = windows[j + 1] - windows[j]
+            result[j] /= span
+    return result
+
+
+def joint_allele_frequency_spectrum(
+        ts, sample_sets, windows=None, polarised=False, mode="site",
+        span_normalise=True):
     """
     Generalised site frequency spectrum.
     """
     method_map = {
-        # "site": site_sample_frequency_spectrum,
-        "branch": branch_sample_frequency_spectrum}
-    return method_map[mode](ts, sample_sets, windows=windows)
+        "site": site_joint_allele_frequency_spectrum,
+        "branch": branch_joint_allele_frequency_spectrum}
+    return method_map[mode](
+        ts, sample_sets, windows=windows, polarised=polarised,
+        span_normalise=span_normalise)
 
 
-class TestSampleFrequencySpectrum(StatsTestCase, SampleSetStatsMixin):
+class TestJointAlleleFrequencySpectrum(StatsTestCase, SampleSetStatsMixin):
 
     # Derived classes define this to get a specific stats mode.
     mode = None
 
     def verify_sample_sets(self, ts, sample_sets, windows):
-        # print("Verify", sample_sets, windows)
-        sfs1 = naive_sample_frequency_spectrum(ts, sample_sets, windows, mode=self.mode)
-        sfs2 = sample_frequency_spectrum(ts, sample_sets, windows, mode=self.mode)
+        # print(ts.genotype_matrix())
+        # print(ts.draw_text())
         windows = ts.parse_windows(windows)
-        self.assertEqual(sfs1.shape[0], len(windows) - 1)
-        self.assertEqual(sfs1.shape, sfs2.shape)
-        # print(sfs1)
-        # print(sfs2)
-        self.assertArrayAlmostEqual(sfs1, sfs2)
-        # print(sfs2.shape)
-
-
-class TestBranchSampleFrequencySpectrum(
-        TestSampleFrequencySpectrum, TopologyExamplesMixin):
+        # TODO implement span_normalise in the low-level code and iterate over this too
```

Comment:
I wondered if we really need span_normalise... but, we do, because if we don't span normalise then the AFS computed in windows of dfferent lengths are not comparable.